pkg/resources/digitalocean/resources.go

```diff
@@ -264,12 +265,14 @@ func deleteVolume(cloud fi.Cloud, t *resources.Resource) error {
 	c := cloud.(do.DOCloud)
 	volume := t.Obj.(godo.Volume)
 	for _, dropletID := range volume.DropletIDs {
-		action, _, err := c.VolumeActionService().DetachByDropletID(context.TODO(), volume.ID, dropletID)
-		if err != nil {
-			return fmt.Errorf("failed to detach volume: %s, err: %s", volume.ID, err)
-		}
-		if err := waitForDetach(c, action); err != nil {
-			return fmt.Errorf("error while waiting for volume %s to detach: %s", volume.ID, err)
+		action, resp, err := c.VolumeActionService().DetachByDropletID(context.TODO(), volume.ID, dropletID)
+		if resp != nil && resp.StatusCode != http.StatusNotFound {
```

Comment:
For the sake of readability and consistency, I'd first check if the error is non-nil and derive proper actions accordingly:

```go
if err != nil {
  if resp != nil && resp.StatusCode == http.StatusNotFound {
    // Volume is already detached, nothing to do.
    continue
  }
  return fmt.Errorf("failed to detach volume %s: %s", volume.ID, err)
}

if err := waitForDetach(c, action); err != nil {
  return fmt.Errorf("error while waiting for volume %s to detach: %s", volume.ID, err)
}
```

Reply:
Sure, updated.